src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentCopy = {};
+    } else if (action.type === 'addProperties') {
+      currentCopy = { ...currentCopy, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentCopy = { ...currentCopy };
+
+      for (const key of action.keysToRemove) {
+        delete currentCopy[key];
+      }
+    }
```

Comment:
There is no default case in your conditional logic. It's a good practice to always include a default case in a switch statement for error handling, even if you are confident that all possible cases are covered.